chalk-solve/src/logging_db.rs

```diff
@@ -0,0 +1,432 @@
+        self.record(impl_id);
+        self.db.borrow().impl_datum(impl_id)
+    }
+
+    fn associated_ty_value(
+        &self,
+        id: crate::rust_ir::AssociatedTyValueId<I>,
+    ) -> Arc<crate::rust_ir::AssociatedTyValue<I>> {
+        let value = self.db.borrow().associated_ty_value(id);
+        self.record(value.impl_id);
+        value
+    }
+
+    fn opaque_ty_data(&self, id: OpaqueTyId<I>) -> Arc<OpaqueTyDatum<I>> {
+        self.record(id);
+        self.db.borrow().opaque_ty_data(id)
+    }
+
+    fn impls_for_trait(
+        &self,
+        trait_id: TraitId<I>,
+        parameters: &[chalk_ir::GenericArg<I>],
+    ) -> Vec<ImplId<I>> {
+        self.record(trait_id);
+        let impl_ids = self.db.borrow().impls_for_trait(trait_id, parameters);
+        self.record_all(impl_ids.iter().copied());
+        impl_ids
+    }
+
+    fn local_impls_to_coherence_check(&self, trait_id: TraitId<I>) -> Vec<ImplId<I>> {
+        self.record(trait_id);
+        self.db.borrow().local_impls_to_coherence_check(trait_id)
+    }
+
+    fn impl_provided_for(&self, auto_trait_id: TraitId<I>, adt_id: AdtId<I>) -> bool {
+        self.record(auto_trait_id);
+        self.record(adt_id);
+        self.db.borrow().impl_provided_for(auto_trait_id, adt_id)
+    }
+
+    fn well_known_trait_id(
+        &self,
+        well_known_trait: crate::rust_ir::WellKnownTrait,
+    ) -> Option<TraitId<I>> {
+        let trait_id = self.db.borrow().well_known_trait_id(well_known_trait);
+        trait_id.map(|id| self.record(id));
+        trait_id
+    }
+
+    fn program_clauses_for_env(
+        &self,
+        environment: &chalk_ir::Environment<I>,
+    ) -> chalk_ir::ProgramClauses<I> {
+        self.db.borrow().program_clauses_for_env(environment)
+    }
+
+    fn interner(&self) -> &I {
+        self.db.borrow().interner()
+    }
+
+    fn trait_name(&self, trait_id: TraitId<I>) -> String {
+        self.db.borrow().trait_name(trait_id)
+    }
+
+    fn adt_name(&self, adt_id: AdtId<I>) -> String {
+        self.db.borrow().adt_name(adt_id)
+    }
+
+    fn assoc_type_name(&self, assoc_ty_id: AssocTypeId<I>) -> String {
+        self.db.borrow().assoc_type_name(assoc_ty_id)
+    }
+
+    fn opaque_type_name(&self, opaque_ty_id: OpaqueTyId<I>) -> String {
+        self.db.borrow().opaque_type_name(opaque_ty_id)
+    }
+
+    fn is_object_safe(&self, trait_id: TraitId<I>) -> bool {
+        self.record(trait_id);
+        self.db.borrow().is_object_safe(trait_id)
+    }
+
+    fn fn_def_datum(&self, _fn_def_id: chalk_ir::FnDefId<I>) -> Arc<FnDefDatum<I>> {
+        todo!("function def datum")
+    }
+}
+
+/// Wraps a [`RustIrDatabase`], and, when dropped, writes out all used
+/// definition to the given file.
+///
+/// Uses [`LoggingRustIrDatabase`] internally.
+///
+/// Uses a separate type, `P`, for the database stored inside to account for
+/// `Arc` or wrapping other storage mediums.
+pub struct WriteOnDropRustIrDatabase<I, W, DB, P = DB>
```

Comment:
Why do we need this? I think it's okay to just make the user output before the LoggingDatabase is dropped manually.

Reply:
In a [previous comment](https://github.com/rust-lang/chalk/pull/430#pullrequestreview-405228281) @nikomatsakis suggested we make the logging database write on drop. We opted to refactor this behavior out into a separate wrapper since we felt that it was distinct, and there are likely use-cases for just logging. Do you think it's worth having?

Reply:
Hmm, I guess it's simple enough to keep.